p4-16/spec/P4-16-spec.mdk

```diff
@@ -7716,33 +7716,54 @@ The evaluation of a P4 program is done in two stages:
   executed to completion, in isolation, when it receives control from the
   architecture
 
-## Compile-time known values { #sec-ct-constants }
+## Compile-time known and local compile-time known values { #sec-compile-time-known }
 
-The following are compile-time known values:
+Certain expressions in a P4 program have the property that their value
+can be determined at compile time. Moreover, for some of these
+expressions, their value can be determined only using information in
+the current scope. We call these _compile-time known values_ and
+_local compile-time known values_ respectively.
+
+The following are local compile-time known values:
 
 - Integer literals, Boolean literals, and string literals.
-- Identifiers declared in an `error`, `enum`, or `match_kind`
-  declaration.
+- Identifiers declared in an `error`, `enum`, or `match_kind` declaration.
 - The `default` identifier.
 - The `size` field of a value with type header stack.
 - The `_` identifier when used as a `select` expression label
-- The expression `{#}` representing an invalid header or header union
-  value.
-- Identifiers that represent declared types, actions, tables, parsers,
-  controls, or packages.
-- Tuple expression where all components are compile-time known values.
-- Structure-valued expressions, where all fields are compile-time
-  known values.
-- Expressions evaluating to a list type, where all elements are
-  compile-time known values.
-- Instances constructed by instance declarations (Section
-  [#sec-instantiations]) and constructor invocations.
-- A legal cast applied to a compile-time known value
-- The following expressions (`+`, `-`, `*`, `/ `, `%`, `!`, `&`, `|`, `&&`, `||`, `<< `, `>> `, `~` `, `\ `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`)
-  when their operands are all compile-time known values.
+- The expression `{#}` representing an invalid header or header union value.
+- Instances constructed by instance declarations (Section [#sec-instantiations]) and constructor invocations.
+- Identifiers that represent declared types, actions, tables, parsers, controls, or packages.
+- Tuple expression where all components are local compile-time known values.
+- Structure-valued expressions, where all fields are local compile-time known values.
+- Expressions evaluating to a list type, where all elements are local compile-time known values.
+- Legal casts applied to local compile-time known values.
+- The following expressions (`+`, `-`, `|+|`, `|-|`, `*`, `/ `, `%`, `!`, `&`, `|`, `^`, `&&`, `||`, `<< `, `>>`, `~`, `/`, `>`, `<`, `==`, `!=`, `<=`, `>=`, `++`, `[:]`, `?:`) when their operands are all local compile-time known values.
+- Expressions of the form `e.minSizeInBits()`, `e.minSizeInBytes()`, `e.maxSizeInBits()` and `e.maxSizeInBytes()` where the type of `e` is not generic.
+
+The following are compile-time known values:
+
+- All local compile-time known values.
+- Constructor parameters.
```

Comment:
I was suprised by this because (probably) I confusing constructor arguments and parameters. I might be worth actually highlighting that here (i.e. that here we are concerned with the values of constructor parameters used in the definition of the control/parser itself).

Reply:
I added " (i.e., the declared parameters for a `parser`, `control`, etc.)"